### PR TITLE
[hail] Shuffler IR TypeCheck

### PR DESCRIPTION
**Pull request by danking**
- add typecheck
- fix shufflergetpartitionbounds type
- add some missing infers

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 1]
```diff
@@ -424,6 +426,31 @@ object TypeCheck {
       case x@WriteValue(value, pathPrefix, spec) =>
         assert(pathPrefix.typ == TString)
       case LiftMeOut(_) =>
+      case ShuffleStart(keyFields, rowType, rowEType, keyEType) =>
+        assert(keyFields.map(_.field).toSet.size == keyFields.length)
+        assert((keyFields.map(_.field).toSet - keyFields.toSet).size == 0)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala:431]: this line is hurting my brain. Is this subtracting a set of a different type?

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 2]
```diff
@@ -424,6 +426,31 @@ object TypeCheck {
...
+      case ShuffleWrite(id, partitionId, rows) =>
+        assert(id.typ == TBinary)
+        assert(partitionId.typ == TInt64)
+        assert(rows.typ.isInstanceOf[TStream])
+      case ShuffleWritingFinished(id, successfulPartitionIds) =>
+        assert(id.typ == TBinary)
+        assert(successfulPartitionIds.typ == TArray(TBinary))
+      case ShuffleGetPartitionBounds(id, nPartitions, keyFields, rowType, keyEType) =>
+        assert(id.typ == TBinary)
+        assert(nPartitions == TInt64)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala:442]: missing `.typ`

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 3]
```diff
@@ -424,6 +426,31 @@ object TypeCheck {
...
+        assert(keyFields.map(_.field).toSet.size == keyFields.length)
+        assert((keyFields.map(_.field).toSet - keyFields.toSet).size == 0)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala:444]: same as above comment on set subtraction

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 4]
```diff
@@ -424,6 +426,31 @@ object TypeCheck {
...
+        val keyType = TArray(rowType.typeAfterSelectNames(keyFields.map(_.field)))
+      case ShuffleRead(id, keyRange, keyFields, rowType, rowEType) =>
+        assert(id.typ == TBinary)
+        assert(keyFields.map(_.field).toSet.size == keyFields.length)
+        assert((keyFields.map(_.field).toSet - keyFields.toSet).size == 0)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala:449]: same

**Comment by danking** [comment]
I'm discovering a number of issues as I try to actually implement these nodes. I'll re-open when that's all sorted.

**Pull request closed without merging**